### ISSUE TITLE
[ENH] - Test for parallel builds

### ISSUE THREAD

**Issue by kcpevey** [conda-incubator/conda-store#548]
### Feature description

We need the ability to build environments in parallel. This functionality was added in https://github.com/conda-incubator/conda-store/issues/392 but there was issues and the work needed to be reverted. 

We need to re-implement this feature. 

### Value and/or benefit

Building all environments in series is not efficient and once there are a number of people on the platform, wait times for builds grows to an unacceptable length of time.

**Comment by kcpevey** [comment]
@costrouc can you help scope this work? Do we need to make some changes to the previous effort or should we take a new direction?

**Comment by costrouc** [comment]
@kcpevey parallel builds are supported. I don't think any action needs to be taken. We just need a release of conda-store.

I'm going to close this issue unless additional work needs to be done.

**Comment by dharhas** [comment]
Have the bugs we found during the pycon tutorials been fixed? Have we tested parallel builds in anyway?

**Comment by kcpevey** [comment]
Are multiple builds of a single environment also parallel? i.e. I click "Create" for an env to be built, immediately realize I forgot a package, edit the yaml, and click "Create" again. 

I've done this quite a few times and the second env immediately says "building" but the time it takes to build seems to indicate its waiting for the first build to finish. This is just a suspicion, since I can't really test or see what's happening as an end user.

**Comment by dharhas** [comment]
reopening until this discussion is resolved.

**Comment by kcpevey** [comment]
Ok, I've been testing on this today. I have a env that takes over an hour to build, call it `slow` and I have an env with just one package that takes one minute to build, call it `fast`. 

Note that when you submit an env to be built, it immediately shows "Building" no matter what. 

What I observed:

1. Build `fast` - took 1 minute
2. Start building `slow`, then start building `fast`. Waited 30 min, `slow` was still building, `fast` was also showing that it was "still" building. Came back even later, both envs were complete. Total build time for `slow` and `fast` was the same (via the admin ui). 
3. Retested steps 1 & 2 and got the same result. 

I do not think parallel builds are working.

**Comment by costrouc** [comment]
@kcpevey what are you testing on. The `main` branch of conda-store does support parallel builds.

I do think the action item here is that we need a test that parallel builds work. I think this feature is also worth mentioning in the docs like you said @dharhas.

**Comment by nkaretnikov** [comment]
I'll investigate more, but here's an experiment I've tried, with some initial results.

Added more logging, to be able to see how long each build takes:
https://github.com/nkaretnikov/conda-store/commits/parallel-builds-548

Created this env:

```yaml
channels:
- conda-forge
dependencies:
- rust>=1.71.1
- binutils-meta>=1.0.4
- gcc>=12.2.0
- llvm-meta>=13.0.1
- cudatoolkit-dev>=11.4.0
description: slow env
name: slow
prefix: null
variables: null
```

Parallel builds are scheduled by clicking on the rebuild button as fast as I can. The logs show that builds start with a few sec delay (limited by how fast I can click on the button -- this is expected).

However, parallel builds are not faster (in total) because each build also takes longer when run.

I would be surprised that this env is big enough to be bound by the resources of the machine I'm using, so it seems there's contention somewhere in our code, maybe the DB (see the lock error below, too).

When scheduling 1 build, one build takes ~2 mins.
When scheduling 4 parallel builds, one build takes ~4 mins.
When scheduling 14 parallel builds, one build takes ~11 mins, with some builds failing due to a DB lock:

```
Traceback (most recent call last):
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/nkaretnikov/conda-store/conda-store-server/conda_store_server/build.py", line 344, in build_conda_docker
    conda_store.container_registry.store_image(
  File "/home/nkaretnikov/conda-store/conda-store-server/conda_store_server/registry.py", line 64, in store_image
    build.id,
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 482, in __get__
    return self.impl.get(state, dict_)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 942, in get
    value = self._fire_loader_callables(state, key, passive)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 973, in _fire_loader_callables
    return state._load_expired(state, passive)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/state.py", line 712, in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/loading.py", line 1451, in load_scalar_attributes
    result = load_on_ident(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/loading.py", line 407, in load_on_ident
    return load_on_pk_identity(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/loading.py", line 530, in load_on_pk_identity
    session.execute(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1714, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1705, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/home/nkaretnikov/.conda/envs/conda-store-dev-old/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: SELECT build.id AS build_id, build.specification_id AS build_specification_id, build.environment_id AS build_environment_id, build.status AS build_status, build.size AS build_size, build.scheduled_on AS build_scheduled_on, build.started_on AS build_started_on, build.ended_on AS build_ended_on, build.deleted_on AS build_deleted_on 
FROM build 
WHERE build.id = ?]
[parameters: (24,)]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

**Comment by nkaretnikov** [comment]
I'll add that my test might be a special case because I'm rebuilding the same env, but it still needs to be looked at.

**Comment by kcpevey** [comment]
>what are you testing on. The main branch of conda-store does support parallel builds.

That's a good point. I'm testing on the nebari instance we have deployed for jatic. We used to be deploying from main, but I dont think that's true for this deployment.

**Comment by nkaretnikov** [comment]
Added a test in https://github.com/conda-incubator/conda-store/pull/567. See the discussion there on potential limitations of this test.

**Comment by nkaretnikov** [comment]
Also, there was this issue regarding DB connections and celery concurrency: https://github.com/conda-incubator/conda-store/issues/247. Because of this, in all example configs, we set the concurrency value to 4, which might not be enough in some environments(?).

But I'm not sure what the current state is, I haven't looked into this DB issue at all.

**Comment by trallard** [comment]
@nkaretnikov  what is the status with this?

**Comment by nkaretnikov** [comment]
@trallard A week+ ago, Chris and I agreed this should be tested by making API requests, so that's the new plan. But I haven't started with this so far. Should I prioritize this over all other things assigned to me? Or is there a particular deadline when you want to see this done?
